### ISSUE TITLE
Postgres | support for NpgsqlRange<T> properties

### ISSUE THREAD

**Issue by crucialize** [linq2db/linq2db#1557]
Linq2db supports NpgsqlRange<T>, but it seems that Properties like NpgsqlRange<T>.LowerBound is not supported.
By the way, is there a plan to support the extension method NpgsqlRange<T>.Contains() ?
```
{Npgsql.PostgresException (0x80004005): 42883: operator does not exist: tsrange > timestamp without time zone
   at Npgsql.NpgsqlConnector.<>c__DisplayClass161_0.<<ReadMessage>g__ReadMessageLong|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlConnector.<>c__DisplayClass161_0.<<ReadMessage>g__ReadMessageLong|0>d.MoveNext() in C:\projects\npgsql\src\Npgsql\NpgsqlConnector.cs:line 1032
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming) in C:\projects\npgsql\src\Npgsql\NpgsqlDataReader.cs:line 444
   at Npgsql.NpgsqlDataReader.NextResult() in C:\projects\npgsql\src\Npgsql\NpgsqlDataReader.cs:line 332
   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken) in C:\projects\npgsql\src\Npgsql\NpgsqlCommand.cs:line 1218
   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior) in C:\projects\npgsql\src\Npgsql\NpgsqlCommand.cs:line 1130
   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader(CommandBehavior behavior)
   at LinqToDB.Data.DataConnection.ExecuteReader(CommandBehavior commandBehavior) in C:\projects\linq2db\Source\LinqToDB\Data\DataConnection.cs:line 1227
   at LinqToDB.Linq.QueryRunner.ExecuteQuery[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Int32 queryNumber)+MoveNext() in C:\projects\linq2db\Source\LinqToDB\Linq\QueryRunner.cs:line 320
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at DouyuGet.Program.Main(String[] args) in ............\Program.cs:line 37}
    Code: "42883"
    ColumnName: null
    ConstraintName: null
    Data: Count = 9
    DataTypeName: null
    Detail: null
    ErrorCode: -2147467259
    File: "parse_oper.c"
    HResult: -2147467259
    HelpLink: null
    Hint: "No operator matches the given name and argument types. You might need to add explicit type casts."
    InnerException: null
    InternalPosition: 0
    InternalQuery: null
    IsTransient: false
    Line: "731"
    Message: "42883: operator does not exist: tsrange > timestamp without time zone"
    MessageText: "operator does not exist: tsrange > timestamp without time zone"
    Position: 135
    Routine: "op_error"
    SchemaName: null
    Severity: "ERROR"
    Source: "Npgsql"
    SqlState: "42883"
    StackTrace: "   at Npgsql.NpgsqlConnector.<>c__DisplayClass161_0.<<ReadMessage>g__ReadMessageLong|0>d.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at Npgsql.NpgsqlConnector.<>c__DisplayClass161_0.<<ReadMessage>g__ReadMessageLong|0>d.MoveNext() in C:\\projects\\npgsql\\src\\Npgsql\\NpgsqlConnector.cs:line 1032\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming) in C:\\projects\\npgsql\\src\\Npgsql\\NpgsqlDataReader.cs:line 444\r\n   at Npgsql.NpgsqlDataReader.NextResult() in C:\\projects\\npgsql\\src\\Npgsql\\NpgsqlDataReader.cs:line 332\r\n   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken) in C:\\projects\\npgsql\\src\\Npgsql\\NpgsqlCommand.cs:line 1218\r\n   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior) in C:\\projects\\npgsql\\src\\Npgsql\\NpgsqlComma
nd.cs:line 1130\r\n   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader(CommandBehavior behavior)\r\n   at LinqToDB.Data.DataConnection.ExecuteReader(CommandBehavior commandBehavior) in C:\\projects\\linq2db\\Source\\LinqToDB\\Data\\DataConnection.cs:line 1227\r\n   at LinqToDB.Linq.QueryRunner.ExecuteQuery[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Int32 queryNumber)+MoveNext() in C:\\projects\\linq2db\\Source\\LinqToDB\\Linq\\QueryRunner.cs:line 320\r\n   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)\r\n   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)\r\n   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)\r\n   at DouyuGet.Program.Main(String[] args) in ..........\\Program.cs:line 37"
    Statement: {SELECT
	p."Id",
	p."ActivityId",
	p."RoomId",
	p."TimeRange",
	p."Content"
FROM
	"public"."Lotteries" p
WHERE
	p."TimeRange" > $1
}
    TableName: null
    TargetSite: {Void MoveNext()}
    Where: null


```

```c#
db.Lotteries.Where(p => ((NpgsqlRange<DateTime>)p.TimeRange).LowerBound > DateTime.Now)
						.ToArray();
```

### Environment details
linq2db version: *2.6.3*
Database Server: *PostgreSQL 11.1 (docker image: postgres:11.1-alpine)*
Database Provider: *Npgsql 4.04*
Operating system: *Windows 10*
Framework version: *.Net core 2.1*

**Comment by sdanyliv** [comment]
You can do it righ now. Define static class
```cs
public static class PgSql
{
     [Sql.Function]
     public static DateTime Lower(NpgsqlRange<DateTime> range)
     {
           return range.LowerBound;
     }
}
```
Then somewhere in initialization define member mapping
```cs
Expressions.MapMember((NpgsqlRange<DateTime> r) => r.LowerBound, r => PgSql.Lower(r));
```

**Comment by crucialize** [comment]
@sdanyliv Thanks for your reply.
I've tried it, and it worked properly.

But it seems that what in the function body doesn't matter. Because this works:
``` c#
	public static class PgSql
	{
		[Sql.Function]
		public static DateTime Lower(NpgsqlRange<DateTime> range)
		{
			throw new NotImplementedException();
		}
	}
```
I guess `[Sql.Function]` and `Lower` tell linq2db to generate `lower(...)` when generating sql string.

But I need to use `.Contains` Method in my project. Fortunately I'm using Linq2db.EFcore, so I can use  the ef context to do something linq2db currently doesn't support.

**Comment by sdanyliv** [comment]
You can use the same technique with Contains or other methods.
If you do not specify method body, yo have to tell about that in attribute `ServerSideOnly = true`.

**Comment by sdanyliv** [comment]
Also check this blog post
http://blog.linq2db.com/

**Comment by sdanyliv** [comment]
@crucialize, looks lik i have to update EFCore integration. I have done several similar mappings before, but providers become smarter and supports more functions.

**Comment by sdanyliv** [comment]
@crucialize, working on automatic support of EF Core functions. It will be delivered with next relaese.

**Comment by MaceWindu** [comment]
by next release you mean 2.6.4 or 2.7?

**Comment by sdanyliv** [comment]
@crucialize, we have released EFCore integration 1.5.0 which should support Npgsql extensions. Package will be available soon.
